pkg/pillar/cmd/zedrouter/flowstats.go

```diff
@@ -0,0 +1,706 @@
+		ipFlow.DstIP = entry.Reverse.SrcIP
+		ipFlow.SrcPort = entry.Forward.SrcPort
+		ipFlow.DstPort = entry.Reverse.DstPort
+		ipFlow.AppInitiate = true
+	} else if forwDstApp {
+		ipFlow.dbg2 = 2
+		ipFlow.DstIP = entry.Forward.DstIP
+		ipFlow.SrcIP = entry.Reverse.DstIP
+		ipFlow.DstPort = entry.Forward.DstPort
+		ipFlow.SrcPort = entry.Reverse.DstPort
+	} else if backSrcApp {
+		ipFlow.dbg2 = 3
+		ipFlow.SrcIP = entry.Forward.SrcIP
+		ipFlow.DstIP = entry.Reverse.SrcIP
+		ipFlow.SrcPort = entry.Forward.SrcPort
+		ipFlow.DstPort = entry.Reverse.SrcPort
+	} else if backDstApp {
+		ipFlow.dbg2 = 4
+		ipFlow.SrcIP = entry.Reverse.DstIP
+		ipFlow.DstIP = entry.Forward.DstIP
+		ipFlow.SrcPort = entry.Reverse.DstPort
+		ipFlow.DstPort = entry.Forward.DstPort
+		ipFlow.AppInitiate = true
+	} else { // if we can not find our App endpoint is part of the flow, something is wrong
+		ipFlow.dbg2 = 5
+		if AppNum > 0 {
+			log.Infof("FlowStats: flow entry can not locate app IP address, appNum %d, %s", AppNum, entry.String())
+		}
+		return ipFlow
+	}
+
+	if ipFlow.AppInitiate {
+		ipFlow.SendPkts = entry.Forward.Packets
+		ipFlow.SendBytes = entry.Forward.Bytes
+		ipFlow.RecvPkts = entry.Reverse.Packets
+		ipFlow.RecvBytes = entry.Reverse.Bytes
+	} else {
+		ipFlow.SendPkts = entry.Reverse.Packets
+		ipFlow.SendBytes = entry.Reverse.Bytes
+		ipFlow.RecvPkts = entry.Forward.Packets
+		ipFlow.RecvBytes = entry.Forward.Bytes
+	}
+	return ipFlow
+}
+
+func checkAppIPAddr(appAddrs []net.IP, entryIP net.IP) bool {
+	for _, appIP := range appAddrs {
+		if strings.Compare(appIP.String(), entryIP.String()) == 0 {
+			return true
+		}
+	}
+	return false
+}
+
+// print the FlowTuple entries
+func (s *flowStats) String() string {
+	tstart := time.Unix(0, int64(s.TimeStart))
+	tout := int32(s.TimeOut)
+	return fmt.Sprintf("TS %v, TO %d(sec), proto %d src=%s dst=%s sport=%d dport=%d, snd=pkts/bytes %d/%d rcv=pkts/bytes %d/%d app-init %v, appNum %d, aclnum %d",
+		tstart, tout, s.Proto, s.SrcIP, s.DstIP, s.SrcPort, s.DstPort, s.SendPkts, s.SendBytes,
+		s.RecvPkts, s.RecvBytes, s.AppInitiate, s.appNum, s.aclNum)
+}
+
+func checkAppAndACL(ctx *zedrouterContext, instData networkAttrs) {
+	pub := ctx.pubAppNetworkStatus
+	items := pub.GetAll()
+	for _, st := range items {
+		status := cast.CastAppNetworkStatus(st)
+		for i, ulStatus := range status.UnderlayNetworkList {
+			log.Infof("===FlowStats: (index %d) AppNum %d, VifInfo %v, IP addr %v, Hostname %s\n",
+				i, status.AppNum, ulStatus.VifInfo, ulStatus.AllocatedIPAddr, ulStatus.HostName)
+
+			if ulStatus.AllocatedIPAddr != "" {
+				tmpIPs := net.ParseIP(ulStatus.AllocatedIPAddr)
+				instData.appIPAddrs[status.AppNum] = append(instData.appIPAddrs[status.AppNum], tmpIPs)
+			}
+
+			ulconfig := ulStatus.UnderlayNetworkConfig
+			intfAttr := aclAttr{
+				vif:     ulStatus.VifInfo.Vif,
+				bridge:  ulStatus.Bridge,
+				netUUID: ulconfig.Network,
+			}
+			instData.bnNet[ulStatus.Bridge] = intfAttr
```

Comment:
This doesn't look right to me. There can be multiple Vifs from different apps connected to a given bridge.

Reply:
right. this bnNet is not binding to app/bridge relationship, it's only getting the bridge specific information for the flow Scope data purpose, so it's each bridge specific. only thing we have app/bridge combination is the aclNum we need to get, that is in ipaclattr

Reply:
i'm overloading the aclAttr structure for two different usage here, mainly is for ipaclattr to get the aclnum. but need to have a bridge indexed structure has all the netUUID, bridge name, etc. So I'm combining them into the same structure. But usage of them is different